npm/parseSelector.go

```diff
@@ -121,73 +121,162 @@ func sortSelector(selector *metav1.LabelSelector) {
 	selector.MatchExpressions = sortedReqs
 }
 
+// getSetNameForMultiValueSelector takes in label with multiple values without operator
+// and returns a new 2nd level ipset name
+func getSetNameForMultiValueSelector(key string, vals []string) string {
+	returnString := key
+	for _, val := range vals {
+		returnString = util.GetIpSetFromLabelKV(returnString, val)
+	}
+	return returnString
+}
+
 // HashSelector returns the hash value of the selector.
 func HashSelector(selector *metav1.LabelSelector) string {
 	sortSelector(selector)
 	return util.Hash(fmt.Sprintf("%v", selector))
 }
 
-// parseSelector takes a LabelSelector and returns a slice of processed labels, keys and values.
-func parseSelector(selector *metav1.LabelSelector) ([]string, []string, []string) {
+// flattenNameSpaceSelector will help flatten multiple NameSpace selector match Expressions values
+// into multiple label selectors helping with the OR condition.
+func FlattenNameSpaceSelector(nsSelector *metav1.LabelSelector) []metav1.LabelSelector {
+	//egress section
+	// for _, egressRule := range nsSelector {
+	// 	log.Logf("%+v", egressRule)
+	// }
+	if nsSelector == nil {
+		return []metav1.LabelSelector{*nsSelector}
+	}
+
+	if len(nsSelector.MatchExpressions) == 0 {
+		return []metav1.LabelSelector{*nsSelector}
+	}
+
+	baseSelector := &metav1.LabelSelector{
+		MatchLabels:      nsSelector.MatchLabels,
+		MatchExpressions: []metav1.LabelSelectorRequirement{},
+	}
+
+	multiValuePresent := false
+	multiValueMatchExprs := []metav1.LabelSelectorRequirement{}
+	for _, req := range nsSelector.MatchExpressions {
+
+		if (req.Operator == metav1.LabelSelectorOpIn) || (req.Operator == metav1.LabelSelectorOpNotIn) {
+			if len(req.Values) == 1 {
+				baseSelector.MatchExpressions = append(baseSelector.MatchExpressions, req)
+				continue
+			}
+			multiValuePresent = true
+			multiValueMatchExprs = append(multiValueMatchExprs, req)
+
+		} else if (req.Operator == metav1.LabelSelectorOpExists) || (req.Operator == metav1.LabelSelectorOpDoesNotExist) {
+			baseSelector.MatchExpressions = append(baseSelector.MatchExpressions, req)
+		} else {
+			log.Errorf("Invalid operator [%s] for selector [%v] requirement", req.Operator, *nsSelector)
+		}
+	}
+
+	// If there are no multiValue NS selector match expressions
+	// return the original NsSelector
+	if !multiValuePresent {
+		return []metav1.LabelSelector{*nsSelector}
+	}
+
+	// Now use the baseSelector and loop over multiValueMatchExprs to create all
```

Comment:
Highly suggest moving codes from 185-192 to "zipMatchExprs". It will improve readability. Now, I need to track for loop in 190 with zipMatchExprs function which uses returned values as input again.

Reply:
Or moving codes from "zipMatchExprs" to here. The function name of "zipMatchExprs" is not very intuitive at least to me.  I think "FlattenNameSpaceSelector" function has a good name and combining all codes in the function is good to understand and not deep.

Reply:
I modelled it as a recursive function, where we take in a slice and expand it using each expr.